File: docs/03-模块设计/datacenter/examples/计算数据点场景.py
# 使用前请在“变量”面板绑定：
# temperature / pressure / setpoint / queryPoint / computePoint / eventPoint
# 调试输入示例：{ "scenario": "read" }

def read_process_snapshot():
    temperature_result = temperature.read()
    pressure_result = pressure.read()
    if temperature_result.code != 0:
        return temperature_result
    if pressure_result.code != 0:
        return pressure_result

    temperature_sample = temperature_result.data
    pressure_sample = pressure_result.data
    if temperature_sample["quality"] != "good" or pressure_sample["quality"] != "good":
        return {
            "valid": False,
            "reason": "输入数据质量不可用",
            "quality": {
                "temperature": temperature_sample["quality"],
                "pressure": pressure_sample["quality"],
            },
        }

    return {
        "valid": True,
        "temperature": temperature_sample["value"],
        "pressure": pressure_sample["value"],
        "load": float(temperature_sample["value"]) * float(pressure_sample["value"]),
        "observedAt": temperature_sample.get("observedAt") or temperature_sample.get("timestamp"),
        "metadata": {
            "path": temperature.path,
            "name": temperature.displayName,
            "dataType": temperature.dataType,
            "unit": temperature.unit,
            "source": temperature.source,
            "attributes": temperature.attributes,
        },
    }

def main(argv, dp, ctx):
    scenario = str(ctx.args.get("scenario") or "read")
    if scenario == "read":
        return read_process_snapshot()
    if scenario == "set":
        target = ctx.args.get("target", argv[0] if len(argv) > 0 else 1200)
        return setpoint.set(float(target))
    if scenario == "execute":
        return queryPoint.execute({"deviceId": ctx.args.get("deviceId", "A01"), "limit": ctx.args.get("limit", 20)})
    if scenario == "run":
        return computePoint.run({"temperature": temperature.get().data, "requestedBy": "compute-demo"})
    if scenario == "publish":
        return eventPoint.publish({"type": "process_snapshot", "payload": read_process_snapshot()})
    if scenario == "unsupported":
        return temperature.history({"limit": 10})
    return {"code": 40032, "msg": "未知 scenario", "data": None}

File: docs/03-模块设计/datacenter/examples/计算数据点场景.js
// 使用前请在“变量”面板绑定：
// temperature / pressure / setpoint / queryPoint / computePoint / eventPoint
// 调试输入示例：{ "scenario": "read" }

function failure(result) {
  return result && result.code !== 0 ? result : null
}

function readProcessSnapshot() {
  const temperatureResult = temperature.read()
  const pressureResult = pressure.read()
  if (failure(temperatureResult)) return temperatureResult
  if (failure(pressureResult)) return pressureResult

  const temperatureSample = temperatureResult.data
  const pressureSample = pressureResult.data
  if (temperatureSample.quality !== 'good' || pressureSample.quality !== 'good') {
    return {
      valid: false,
      reason: '输入数据质量不可用',
      quality: {
        temperature: temperatureSample.quality,
        pressure: pressureSample.quality,
      },
    }
  }

  return {
    valid: true,
    temperature: temperatureSample.value,
    pressure: pressureSample.value,
    load: Number(temperatureSample.value) * Number(pressureSample.value),
    observedAt: temperatureSample.observedAt || temperatureSample.timestamp,
    metadata: {
      path: temperature.path,
      name: temperature.displayName,
      dataType: temperature.dataType,
      unit: temperature.unit,
      source: temperature.source,
      attributes: temperature.attributes,
    },
  }
}

function previewSetpoint() {
  const target = Number(ctx.args.target ?? argv[0] ?? 1200)
  if (!Number.isFinite(target)) return { code: 40032, msg: 'target 必须是数字', data: null }
  return setpoint.set(target)
}

function previewQueryExecution() {
  return queryPoint.execute({
    deviceId: ctx.args.deviceId || 'A01',
    limit: Number(ctx.args.limit || 20),
  })
}

function previewComputeRun() {
  const current = temperature.get()
  if (failure(current)) return current
  return computePoint.run({
    temperature: current.data.value,
    requestedBy: 'compute-demo',
  })
}

function previewPublish() {
  return eventPoint.publish({
    type: 'process_snapshot',
    payload: readProcessSnapshot(),
    emittedAt: new Date().toISOString(),
  })
}

switch (String(ctx.args.scenario || 'read')) {
  case 'read':
    return readProcessSnapshot()
  case 'set':
    return previewSetpoint()
  case 'execute':
    return previewQueryExecution()
  case 'run':
    return previewComputeRun()
  case 'publish':
    return previewPublish()
  case 'unsupported':
    return temperature.history({ limit: 10 })
  default:
    return { code: 40032, msg: '未知 scenario', data: null }
}
